### PR TITLE
Repl & Debugger

### PR DESCRIPTION
**Pull request by kustosz**
### Pull Request Description
Adds a REPL functionality and a debugger that can be triggered from inside Enso code.

### Important Notes
No trickery this time, should be self-explanatory.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/Repl.scala [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+/**
+  * The Repl logic to inject into runtime instrumentation framework.
+  *
+  * @param replIO the IO implementation to use with this Repl
+  */
+case class Repl(replIO: ReplIO) extends ReplDebuggerInstrument.SessionManager {
+
+  /**
+    * Runs the Repl session by asking for user input and performing
```
Comment (iamrecursion) [inline engine/language-server/src/main/scala/org/enso/languageserver/Repl.scala]: I would note here that end of input kills the repl.

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/ReplDebuggerInstrument.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        EventContext eventContext, SessionManagerReference sessionManagerReference) {
+      this.eventContext = eventContext;
+      this.sessionManagerReference = sessionManagerReference;
+    }
+
+    private Object getValue(MaterializedFrame frame, FramePointer ptr) {
+      return getProperFrame(frame, ptr).getValue(ptr.getFrameSlot());
+    }
+
+    private MaterializedFrame getProperFrame(MaterializedFrame frame, FramePointer ptr) {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/instrument/ReplDebuggerInstrument.java]: I feel like I've seen this logic somewhere before. Is it a duplicate?

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java [hunk 3]
```diff
@@ -20,14 +20,28 @@
 /** Node running Enso expressions passed to it as strings. */
 @NodeInfo(description = "Evaluates code passed to it as string")
 public abstract class EvalNode extends BaseNode {
```
Comment (iamrecursion) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java]: Is there a reason this isn't under `expression/builtin` in the source tree?

File: engine/runtime/src/main/java/org/enso/interpreter/instrument/ReplDebuggerInstrument.java (hunk 2 above)
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/instrument/ReplDebuggerInstrument.java]: Duplicate-ish. It's very close to the logic of reading a local variable in the relevant node. Slightly different (simpler) though, so I went with a non-exact duplicate for now.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java (hunk 3 above)
Comment (kustosz) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/debug/EvalNode.java]: `expression/builtin` is for builtin function root nodes. This one is not a root node, wrapped in its counterpart inside `expression.builtin.debug.DebugEvalNode`

**Pull request merged**